### PR TITLE
amp instagram converter

### PR DESCRIPTION
**Pull request by mgonzo**
This converter is written against the tests. I have not been able to test it in the sheknows project because I have not been able to find an article yet that uses the instagram embed.

TODO:
Test in article body that uses instagram embed.

### CODE REVIEW

File: src/Converter/Extensions/InstagramConverter.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+use League\Event\EventInterface;
+use Predmond\HtmlToAmp\ElementInterface;
+use Predmond\HtmlToAmp\Converter\ConverterInterface;
+
+class InstagramConverter implements ConverterInterface
+{
+    public function hasInstagram(ElementInterface $element)
+    {
+        $classes = explode(' ', $element->getAttribute('class'));
+        return in_array('instagram-media', $classes);
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php]: Can you also add a check for the attribute `data-instgrm-version` as well? Both checks are good, and the `data-instgrm-version` seems to be a staple. You don't have to check the value, but you could do something like this:

``` php
array_key_exists('data-instgrm-version', $element->getAttributes());
```

File: src/Converter/Extensions/InstagramConverter.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            'height' => 384,
+            'data-shortcode' => $shortcode
+        ];
+
+        return $element->createWritableElement('amp-instagram', $attrs);
+    }
+
+    public function handleInstagram( EventInterface $event, ElementInterface $element) {
+
+        if ($this->hasInstagram($element) == false) {
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php]: Instead of `hasInstagram` I prefer `shouldBeConverted($element)` because it is more clear the intention. You can combine these two `if` statements and nest the `hasChildren()` check in the `shouldBeConverted()` method:

``` php
if ($this->shouldBeConverted() === false) {
    return;
}
```

File: src/Converter/Extensions/InstagramConverter.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+            return;
+        }
+
+        if ($element->hasChildren() == false) {
+            return;
+        }
+
+        $event->stopPropagation();
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php:63]: You shouldn't stop propagation until you know that you were able to parse a shortcode.

File: src/Converter/Extensions/InstagramConverter.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        $shortcode = $this->findTheCode($element);
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php]: I prefer `$this->getEmbedId($element)` or `$this->getEmbedShortcode($element)`.

File: src/Converter/Extensions/InstagramConverter.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        $element->replaceWith($this->getAmpInstagram($element, $shortcode));
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php]: Once you know you have a valid embed ID, then (and only then) call replace. In this case `getAmpInstagram()` is an over-optimization and you can just create the writable element in the body of this handler.

File: src/Converter/Extensions/InstagramConverter.php [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        $classes = explode(' ', $element->getAttribute('class'));
+        return in_array('instagram-media', $classes);
+    }
+
+    public function findTheCode($element) 
+    {
+        $href = $element->getAttribute('href');
+        if ($href != '') {
+            $parts = explode('/', $href);
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php]: I think a regex would be more consistent here:

``` php
if (1 === preg_match('/(?:instagr\.am|instagram\.com)\/p\/([^\/]+)\/?$/i', $href, $matches)) {
    return $matches[1];
}
```

See this demo http://regexr.com/3cov4 and see this https://github.com/xentek/instagram-embed/blob/master/instagram-embed.php#L20

File: spec/Converter/Extensions/InstagramConverterSpec.php [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        $this->getAmpInstagram($element, $code)->shouldReturnAnInstanceOf(\Predmond\HtmlToAmp\Element::class);
+    }
+
+    public function it_replaces_element_class_instagram_with_amp(
+        EventInterface $event,
+        ElementInterface $element
+    ) {
+        $this->handleInstagram($event, $element);
+    }
+
```
Comment (paulredmond) [inline spec/Converter/Extensions/InstagramConverterSpec.php:62]: Missing a spec for `getSubscribedEvents()`. See https://github.com/paulredmond/html-to-amp/blob/master/spec/Converter/ImageConverterSpec.php#L50-L55

File: spec/Converter/Extensions/InstagramConverterSpec.php [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            'layout' => "responsive",
+            'width' => 600,
+            'height' => 384,
+            'data-shortcode' => 'BBB-bbb'
+        ];
+
+        $code = 'BBB-bbb';
+
+        $element->createWritableElement('amp-instagram', $attrs)->willReturn($newElement);
+        $this->getAmpInstagram($element, $code)->shouldReturnAnInstanceOf(\Predmond\HtmlToAmp\Element::class);
```
Comment (paulredmond) [inline spec/Converter/Extensions/InstagramConverterSpec.php]: We need a test for when things go bad, like `null` is returned, and anywhere the logic can splinter.

File: src/Converter/Extensions/InstagramConverter.php [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+{
+    public function hasInstagram(ElementInterface $element)
+    {
+        $classes = explode(' ', $element->getAttribute('class'));
+        return in_array('instagram-media', $classes);
+    }
+
+    public function findTheCode($element) 
+    {
+        $href = $element->getAttribute('href');
```
Comment (paulredmond) [inline src/Converter/Extensions/InstagramConverter.php:25]: As mentioned lower, I prefer `$this->getEmbedId($element)` or `$this->getEmbedShortcode($element)` for this method name.

**Comment by paulredmond** [comment]
@mgonzo looks good, I've made some suggestions :+1:

**Comment by AMPtemplates** [comment]
@mgonzo: nice work indeed and great idea ;) we are trying the same.  Please check and comment on this 0.1 version at http://amphtmlconverter.com

**Comment by paulredmond** [comment]
:+1:
